client/ayon_silhouette/api/lib.py

```diff
@@ -0,0 +1,243 @@
+        fx.select(previous_selection)
+        pass
+
+
+@contextlib.contextmanager
+def undo_chunk(label=""):
+    """Open undo chunk during context.
+
+    In Silhouette, it's often necessary to group operations into an undo chunk
+    to ensure the UI updates correctly on property and value changes.
+
+    Note that `contextlib.contextmanager` can also be used as function
+    decorators.
+
+    """
+    try:
+        fx.beginUndo(label)
+        yield
+    finally:
+        fx.endUndo()
+
+
+def imprint(node, data: Optional[dict], key="AYON"):
+    """Write `data` to `node` as userDefined attributes
+
+    Arguments:
+        node (fx.Object | fx.Node): The selection object
+        data (dict): Dictionary of key/value pairs
+    """
+
+    # Remove data
+    if data is None:
+        if isinstance(node, fx.Node):
+            node.setState(key, None)
+        else:
+            node.removeProperty(key)
+        return
+
+    # Set data
+    if isinstance(node, fx.Node):
+        node.setState(key, data)
+    elif isinstance(node, fx.Object):
+        # Serialize data
+        value = json.dumps(data)
+        prop = node.property(key)
+        if not prop:
+            # Create property
+            # Arguments are `id`, `label` and `default_value`. By passing the
+            # default value as empty string we make it a string attribute.
+            prop = fx.Property(key, key, "")
+            prop.hidden = True
+            node.addProperty(prop)
+
+        # Set value
+        prop.value = value
+    else:
+        raise TypeError(f"Unsupported node type: {node} ({type(node)})")
+
+
+def read(node, key="AYON") -> Optional[dict]:
+    """Return user-defined attributes from `node`
+
+    Arguments:
+        node (fx.Object | fx.Node): Node or object to redad from.
+        key (str): The key to read from.
+
+    Returns:
+        Optional[dict]: The data stored in the node.
+
+    """
+    if isinstance(node, fx.Node):
+        # Use node state instead of property
+        return node.getState(key)
+    elif isinstance(node, fx.Object):
+        # Project or source items do not have state
+        prop = node.property(key)
+        if not prop:
+            return
+        value = prop.value
+        if not value:
+            return
+
+        try:
+            return json.loads(value)
+        except json.JSONDecodeError as exc:
+            log.error(
+                f"Failed to read '{key}' from node {node}"
+                f" with value {value}: {exc}")
+            return
+    else:
+        raise TypeError(f"Unsupported node type: {node} ({type(node)})")
+
+
+def set_new_node_position(node):
+    """Position the node near the active node, or the top-right of the scene"""
+    n = fx.activeNode()
+    if n:
+        pos = fx.trees.nextPos(n)
+    else:
+        bounds = fx.trees.bounds
+        size = fx.trees.nodeSize(node)
+        pos = fx.Point(bounds.right - size.x / 2,
+                    bounds.top + size.y / 2)
```

Comment:
```suggestion
        pos = fx.Point(
            bounds.right - size.x / 2,
            bounds.top + size.y / 2
        )
```

Reply:
This is basically copied from an example action / source code of Silhouette itself. I didn't bother confirming whether there code works as intended haha 🗡️

Reply:
The suggestion was just formatting change, the comment below was additional question.